File: Edu-CirlceUp/services/translation_service.py
"""
Translation Service
Primary:  Helsinki-NLP MarianMT models (HuggingFace — offline, free)
Fallback: LibreTranslate (Docker service)
"""

import logging
import requests
from functools import lru_cache

from config import get_settings

logger   = logging.getLogger(__name__)
settings = get_settings()

LANGUAGE_CODE_MAP = {
    "hindi":      "hi",
    "tamil":      "ta",
    "telugu":     "te",
    "bengali":    "bn",
    "marathi":    "mr",
    "gujarati":   "gu",
    "kannada":    "kn",
    "malayalam":  "ml",
    "punjabi":    "pa",
    "urdu":       "ur",
    "french":     "fr",
    "german":     "de",
    "spanish":    "es",
    "arabic":     "ar",
    "chinese":    "zh",
    "japanese":   "ja",
    "korean":     "ko",
    "russian":    "ru",
}


@lru_cache(maxsize=8)
def _load_helsinki_pipeline(target_code: str):
    """Load MarianMT model — cached so it loads only once per language."""
    from transformers import pipeline
    model_name = f"Helsinki-NLP/opus-mt-en-{target_code}"
    try:
        return pipeline("translation", model=model_name)
    except Exception as e:
        logger.warning("Helsinki model not available for %s: %s", target_code, e)
        return None


def translate_text(text: str, target_language: str) -> str:
    """
    Translate English text to target language.
    Tries Helsinki-NLP first, falls back to LibreTranslate.
    """
    if not text or not text.strip():
        return text

    target_code = LANGUAGE_CODE_MAP.get(target_language.lower())
    if not target_code:
        logger.warning("Unsupported language: %s", target_language)
        return text

    # Try Helsinki-NLP (offline, free)
    try:
        pipe = _load_helsinki_pipeline(target_code)
        if pipe:
            result = pipe(text[:1000])  # MarianMT limit
            return result[0]["translation_text"]
    except Exception as e:
        logger.warning("Helsinki translation failed: %s", e)

    # Fallback: LibreTranslate
    return _libretranslate(text, target_code)


def _libretranslate(text: str, target_code: str) -> str:
    """Call local LibreTranslate container."""
    try:
        resp = requests.post(
            f"{settings.libre_translate_url}/translate",
            json={"q": text, "source": "en", "target": target_code, "format": "text"},
            timeout=15
        )
        if resp.ok:
            return resp.json().get("translatedText", text)
    except Exception as e:
        logger.error("LibreTranslate failed: %s", e)
    return text


def supported_languages() -> list:
    return list(LANGUAGE_CODE_MAP.keys())

File: Langchain/src/main/java/com/EduCircle/Langchain/Service/TranslationService.java
package com.EduCircle.Langchain.Service;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.stereotype.Service;
import org.springframework.web.reactive.function.client.WebClient;

import java.util.List;
import java.util.Map;

@Service
@RequiredArgsConstructor
@Slf4j
public class TranslationService {

    private final WebClient aiWebClint;

    private static final List<String>SUPPORTED_LANGUAGES=List.of(
            "hindi", "tamil", "telugu", "bengali", "marathi", "gujarati",
            "kannada", "malayalam", "punjabi", "urdu",
            "french", "german", "spanish", "arabic",
            "chinese", "japanese", "korean", "russian"
    );

    @Cacheable(value = "translations",key ="#text.hashCode() + '_' + #targetLanguage" )
    public String translate(String text,String targetLanguage){
        if(targetLanguage==null || targetLanguage.equalsIgnoreCase("english")){
            return text;
        }
        try {
            Map<String,Object>response=aiWebClint.post().uri("/translate").bodyValue(Map.of(
                    "text",            text,
                    "target_language", targetLanguage
            )).retrieve().bodyToMono(Map.class).block();
            if (response!=null && response.containsKey("translated")){
                return response.get("translated").toString();
            }
        }
        catch (Exception e){
            log.error("Translation failed for language {} : {} ",targetLanguage,e.getMessage());
        }
        return text;
    }
    public List<String> getSupportedLanguages() {
        return SUPPORTED_LANGUAGES;
    }
}
